### PR TITLE
Change `create_task` type

### LINKED ISSUE #6770: asyncio.create_task hints say you can pass any Awaitable, but you cannot pass asyncio.Tasks
The following program passes mypy type checking, but fails at runtime:

```python
import asyncio
from typing import Awaitable

async def main() -> None:
    async def coro() -> Awaitable[None]: pass
    task = asyncio.create_task(coro())
    asyncio.create_task(task)

asyncio.run(main())
```

```
(venv) ➜ python --version
Python 3.9.6
(venv) ➜ mypy --version
mypy 0.930
(venv) ➜ mypy --strict example.py 
Success: no issues found in 1 source file
(venv) ➜ python example.py
Traceback (most recent call last):
  File ".../create-task-types/example.py", line 10, in <module>
    asyncio.run(main())
  File ".../python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File ".../python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File ".../create-task-types/example.py", line 7, in main
    asyncio.create_task(task)
  File ".../python3.9/asyncio/tasks.py", line 361, in create_task
    task = loop.create_task(coro)
  File ".../python3.9/asyncio/base_events.py", line 433, in create_task
    task = tasks.Task(coro, loop=self, name=name)
TypeError: a coroutine was expected, got <Task pending name='Task-2' coro=<main.<locals>.coro() running at .../create-task-types/example.py:4>>
```

### PR DESCRIPTION
**Pull request by sobolevn**
I've manually tested that `Awaitable`s are not supported:

```python
import asyncio

class My:
    def __await__(self):
        async def inner():
            print('here')
        return inner

async def main():
  await asyncio.create_task(My())

asyncio.run(main())
```

Output:

```
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython/build/ex.py", line 15, in <module>
    asyncio.run(main())
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/sobolev/Desktop/cpython/build/ex.py", line 13, in main
    await asyncio.create_task(My())
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/asyncio/tasks.py", line 361, in create_task
    task = loop.create_task(coro)
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/asyncio/base_events.py", line 433, in create_task
    task = tasks.Task(coro, loop=self, name=name)
TypeError: a coroutine was expected, got <__main__.My object at 0x10c38fe20>
```

But, `Generator` objects are also supported:

```python
import asyncio
def gen():
   i = yield
   print(1, i)  # prints: `1 None`

async def main():
  await asyncio.create_task(gen())

asyncio.run(main())
```

Closes #6770

### CODE REVIEW

File: stdlib/asyncio/tasks.pyi
Comment (sobolevn) [inline stdlib/asyncio/tasks.pyi:294]: Notice that `_TaskYieldType` was not used.

**Comment by sobolevn** [comment]
Ok, we have some problems:

```python
self._navigationPromise = self._loop.create_task(asyncio.wait([
    self._lifecycleCompletePromise,
    self._createTimeoutPromise(),
], return_when=concurrent.futures.FIRST_COMPLETED))
```

Here `asyncio.wait` returns `Future`, but now `self._loop.create_task()` wants `Coroutine`. `Future` is not compatible with `Coroutine`. 🤔

File: stdlib/asyncio/tasks.pyi
Comment (Akuli) [reply stdlib/asyncio/tasks.pyi:294]: This doesn't seem right. Event loops typically use yields to communicate between tasks and the loop, so the generator must yield objects of a specific type.

**Comment by Akuli** [comment]
Doesn't look like a `Future` to me:

```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> asyncio.wait([])
<coroutine object wait at 0x7f90e8cef140>
```

File: stdlib/asyncio/tasks.pyi
Comment (sobolevn) [reply stdlib/asyncio/tasks.pyi:294]: Some experiments:

```python
import asyncio
def gen():
   i = yield
   print(1, i)  # prints: `1, None`
   return 'a'

async def main():
  print('res', await asyncio.create_task(gen()))  # prints: `res, a`

asyncio.run(main())
```

And:

```python
import asyncio
def gen():
   i = yield
   print(1, i)  # prints: `1, None`
   yield 'a'

async def main():
  print('res', await asyncio.create_task(gen()))

asyncio.run(main())
```

```
1 None
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython/build/ex.py", line 10, in <module>
    asyncio.run(main())
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/sobolev/.pyenv/versions/3.9.9/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/sobolev/Desktop/cpython/build/ex.py", line 8, in main
    print('res', await asyncio.create_task(gen()))  # prints: `res, a`
  File "/Users/sobolev/Desktop/cpython/build/ex.py", line 5, in gen
    yield 'a'
RuntimeError: Task got bad yield: 'a'
```

**Comment by sobolevn** [comment]
@Akuli sorry, I had to clarify that I talking about typeshed's definitions: https://github.com/python/typeshed/blob/505ea726415016e53638c8b584b8fdc9c722cac1/stdlib/asyncio/tasks.pyi#L235-L239

**Comment by Akuli** [comment]
A few observations:
- The `coro` argument gets passed to `Task.__init__`, so that should be changed too.
- `Task.__init__` uses `asyncio.coroutines.iscoroutine`, which is an optimized `isinstance(obj, asyncio.coroutines._COROUTINE_TYPES)`, not to be confused with `inspect.iscoroutine` which is `isinstance(obj, types.CoroutineType)`.
- asyncio is a mess :(

**Comment by sobolevn** [comment]
Ok, I have no idea what is going on:
- Why some definitions are `async def`, while others are sync `def` that return `Future`?
- Why does it return `Future`, why not `Coroutine`?
- Why `stubtest` allows this?

**Comment by sobolevn** [comment]
Ok, my plan is to sync all `async def` in source code and `async def` in `typeshed` (unless there are some problems that I don't know about).

**Comment by hauntsaninja** [comment]
Probably shouldn't merge this as is, given the false positives. The one in my code could be potentially worked around by changing `AsyncIterator.__anext__` to return a `Coroutine`.

Re stubtest:
It's one of many tricks stubtest doesn't know about / mypy itself largely treats `def f() -> Coroutine[Any, Any, int]` the same as `async def f() -> int` so it's not something you're forced to handle. Can't remember if I tried adding support for this to stubtest; there were several things I chose not to add based on typeshed error counts. Might have been affected by older Pythons, e.g. on Python 3.5 several functions are `@coroutine` not `async def`.

**Comment by sobolevn** [comment]
Totally do not merge this as-is, I will spend several days fixing all the `async def` / `def` stuff. And hopefully after that we can return to this issue.

**Comment by srittau** [comment]
I've converted this to a draft for now. Please mark as "ready for review" when it is. :)

**Comment by sobolevn** [comment]
I've changed all things I could find to be `async def`: https://github.com/python/typeshed/blob/7bc9c1625eb0df0500e70de96f3e1a698bb8c05a/stdlib/asyncio/tasks.pyi#L235-L239

**Comment by sobolevn** [comment]
Analysis:
- `homeasistant` annotation is not correct, it uses `Awaitable`, which is not supported by `create_task`: https://github.com/home-assistant/core/blob/19b7454161046d68dc388e7ff1ffdd0573855449/homeassistant/core.py#L439 I will send a PR with the fix to their repo 👍 
- @hauntsaninja, what do you think about https://github.com/hauntsaninja/boostedblob/blob/master/boostedblob/boost.py#L547 ?
- `Argument 1 to "create_task" has incompatible type "Awaitable[Sequence[Any]]"; expected "Union[Generator[Any, None, <nothing>], Coroutine[Any, Any, <nothing>]]"` seems correct to me

**Comment by srittau** [comment]
@sobolevn If I understand correctly, in your opinion the primer diffs all show genuine problems?

**Comment by srittau** [comment]
Ok, I'll have a look once the merge conflicts are resolved.

File: stdlib/asyncio/base_events.pyi
Comment (srittau) [inline stdlib/asyncio/base_events.pyi:12]: Nit for future reference: `Coroutine` should now be imported from `collections.abc`.

**Review by srittau — APPROVED** [review]

**Comment by sobolevn** [comment]
Thanks everyone 🎉

**Pull request merged**
